CLUSTER.md

```diff
@@ -0,0 +1,490 @@
+with the Failover Coordinator and its Primary. In case the node is a Primary which hasn't been demoted by the admin, it will
+stop responding to client requests but keep replication connection open to prevent data loss (since it will no longer succeed
+in emitting a heartbeat, fail-over will soon take place).
+
+#### Migrate a Shard between Failover Coordinators
+
+*One of the Flotilla invariants is that all nodes in a certain shard belong to the same Failover Coordinator -- this means we must
+move an entire shard at once, and not one by one. In order to move a single node between Failover Coordinators one must first
+remove the association with its shard*
+
+When the Topology Director receives a request to move a shard to Failover Coordinator X, it simply sets the correct Failover Coordinator in its topology and increases
+the epoch. This new topology will then propagate to the various Failover Coordinators and nodes. There's no requirement for ordering in this
+flow -- either the target Failover Coordinator or the source Failover Coordinator can get the message first and the system will still achieve the desired end
+state.
+When an Failover Coordinator receives a heartbeat request from a node it does not own it returns a MOVED response to the correct Failover Coordinator
+(there will always be such an Failover Coordinator since the replacement is done in an atomic manner). As an implementation detail, after
+being moved between Failover Coordinators a node should wait double its normal time before initiating failover in order to avoid redundant
+fail-overs.
+If the target Failover Coordinator is still unaware of the new topology it might cause the shards to ping-pong a bit between the Failover Coordinators, so a node should contain logic to reduce
+the frequency of heartbeats after a moved response, but even if it doesn't this situation will self-resolve without
+causing any damage.
+
+Worth noting that this flow is the main reason why a Topology Director **must** ensure no concurrent updates to the same shard:
+- Shard X comprised of 3 nodes X1,X2,X3 belong to Failover Coordinator A
+- X is moved to Failover Coordinator B
+- All entities other than X3 get the updated topology
+- X1 sends a heartbeat as primary to Failover Coordinator **B**
+- X is moved back to Failover Coordinator A, but Failover Coordinator B is delayed in receiving the updated topology
+- X3 sends a heartbeat as primary to Failover Coordinator **A**
+While this situation will self-resolve when Failover Coordinator B receives the updated topology, Flotilla aims to guarantee a strongly 
+consistent topology - so it needs to prevent concurrent updates.
+
+![Move Shard](images/move_shards.jpg?raw=true "Move shard between Failover Coordinators")
+
+### Failover
+
+Every K milliseconds, every node sends a heartbeat to the Failover Coordinator which owns it. This heartbeat includes:
```

Comment:
There are many issues related to cluster failover due to long running commands (LUA scripts, large sets joins etc...) I wonder if we want to improve the heartbeat mechanism to overcome these issues. Maybe this is a separate discussion but I think that redesign the cluster should allow users to run long commands without having to keep adjusting the timeouts values.

Reply:
This is something I've mentioned before and want to address. I want the heartbeating mechanism to be pulled off the main thread of Redis so we can heartbeat to the FCs inside of commands that take single digit seconds.

Reply:
It's possible to do, and won't impact the rest of the design -- but is it always desirable? I'd argue that after a certain threshold we _do_ want to failover a primary with a non-responsive main thread.

The ideal solution would probably be to have a heartbeat thread which also monitors the main one and can decide to stop heartbeating if the main thread is unresponsive for a long enough time.

Reply:
> It's possible to do, and won't impact the rest of the design

100% correct, it's an implementation decision.

> but is it always desirable?
I think so, we've had quite a bit of issue reliably determining the state of the execution thread in Redis since it is synchronous. 

>  I'd argue that after a certain threshold we do want to failover a primary with a non-responsive main thread.  I'd argue that after a certain threshold we do want to failover a primary with a non-responsive main thread.

This is mostly what I'm expecting. I suppose we would have two thresholds.
* No heartbeat received after X seconds: The process is bricked for some reason and we should declare it unhealthy so failover can happen.
* Too many heartbeats of the same type received for Y seconds: Something is stuck and not making progress fast enough, maybe a sync flush was executed. We should still failover eventually, but give some more room here.

Reply:
Conceptually speaking, failure detection is orthogonal to the cluster V2 spec discussion so I agree it can wait or be explored in parallel.

I think the key question in the domain of failure detection is "progress", i.e., what do we mean by "progress", especially when a script is involved? It is not a trivial problem IMO to 100% reliably determine if an arbitrary command is running in an infinite loop, or in other words, making "progress", or not. Therefore, I think there should always be an upper bound on how long a command can run. Whether or not we need a separate thread to pump h/b is a secondary question to me as I think the main thread is the best place to implement the logic to determine "progress". The separate thread is more like the IO threads, which can pick up the "progress" breadcrumb left by the main thread and report it with the h/b. It is just there to help reduce the work on the main thread but the main thread could just do it on its own as well.